File: ood/16_facebook.py
#!/usr/bin/env python3
# coding: utf-8



class ConnectionInvitationStatus(Enum):
  PENDING, ACCEPTED, REJECTED, CANCELED = 1, 2, 3, 4


class AccountStatus(Enum):
  ACTIVE, CLOSED, CANCELED, BLACKLISTED, DISABLED = 1, 2, 3, 4, 5


class Address:
  def __init__(self, street, city, state, zip_code, country):
    self.__street_address = street
    self.__city = city
    self.__state = state
    self.__zip_code = zip_code
    self.__country = country




# For simplicity, we are not defining getter and setter functions. The reader can
# assume that all class attributes are private and accessed through their respective
# public getter methods and modified only through their public methods function.

class Account:
  def __init__(self, id, password, status=AccountStatus.Active):
    self.__id = id
    self.__password = password
    self.__status = status

  def reset_password(self):
    None


# from abc import ABC, abstractmethod
class Person(ABC):
  def __init__(self, name, address, email, phone, account):
    self.__name = name
    self.__address = address
    self.__email = email
    self.__phone = phone
    self.__account = account


class Member(Person):
  def __init__(self, id, date_of_membership, name):
    self.__member_id = id
    self.__date_of_membership = date_of_membership
    self.__name = name

    self.__profile = Profile()
    self.__member_follows = []
    self.__member_connections = []
    self.__page_follows = []
    self.__member_suggestions = []
    self.__connection_invitations = []
    self.__group_follows = []

  def send_message(self, message):
    None

  def create_post(self, post):
    None

  def send_connection_invitation(self, invitation):
    None

  def search_member_suggestions(self):
    None


class Admin(Person):
  def block_user(self, customer):
    None

  def unblock_user(self, customer):
    None

  def enable_page(self, page):
    None

  def disable_page(self, page):
    None


class ConnectionInvitation:
  def __init__(self, member_invited, name, status=ConnectionInvitationStatus.PENDING):
    self.__member_invited = member_invited
    self.__status = status
    self.__date_created = datetime.date.today()
    self.__date_updated = datetime.date.today()

  def accept_connection(self):
    None

  def reject_connection(self):
    None




class Profile:
  def __init__(self, profile_picture, cover_photo, gender):
    self.__profile_picture = profile_picture
    self.__cover_photo = cover_photo
    self.__gender = gender

    self.__work_experiences = []
    self.__educations = []
    self.__places = []
    self.__stats = []

  def add_work_experience(self, work):
    None

  def add_education(self, education):
    None

  def add_place(self, place):
    None


class Work:
  def __init__(self, title, company, location, date_from, date_to, description):
    self.__title = title
    self.__company = company
    self.__location = location
    self.__from = date_from
    self.__to = date_to
    self.__description = description





class Page:
  def __init__(self, id, name, description, type, total_members):
    self.__page_id = id
    self.__name = name
    self.__description = description
    self.__type = type
    self.__total_members = total_members
    self.__recommendation = []

  def get_recommendation(self):
    return self.__recommendation


class Recommendation:
  def __init__(self, id, rating, description):
    self.__recommendation_id = id
    self.__rating = rating
    self.__description = description
    self.__created_at = datetime.date.today()






class Group:
  def __init__(self, id, name, description, total_members):
    self.__group_id = id
    self.__name = name
    self.__description = description
    self.__total_members = total_members
    self.__members = []

  def add_member(self, member):
    None

  def update_description(self, description):
    None


class Post:
  def __init__(self, id, text, total_likes, total_shares, owner):
    self.__post_id = id
    self.__text = text
    self.__total_likes = total_likes
    self.__total_shares = total_shares
    self.__owner = owner


class Message:
  def __init__(self, id, sent_to, body, media):
    self.__message_id = id
    self.__sent_to = sent_to
    self.__message_body = body
    self.__media = media

  def add_member(self, member):
    None


class Comment:
  def __init__(self, id, text, total_likes, owner):
    self.__comment_id = id
    self.__text = text
    self.__total_likes = total_likes
    self.__owner = owner






from abc import ABC, abstractmethod

class Search(ABC):
  def search_member(self, name):
    None

  def search_group(self, name):
    None

  def search_page(self, name):
    None

  def search_post(self, word):
    None


class SearchIndex(Search):
  def __init__(self):
    self.__member_names = {}
    self.__group_names = {}
    self.__page_titles = {}
    self.__posts = {}

  def add_member(self, member):
    if member.get_name() in self.__member_names:
      self.__member_names.get(member.get_name()).add(member)
    else:
      self.__member_names[member.get_name()] = member

  def add_group(self, group):
    None

  def add_page(self, page):
    None

  def add_post(self, post):
    None

  def search_member(self, name):
    return self.__member_names.get(name)

  def search_group(self, name):
    return self.__group_names.get(name)

  def search_page(self, name):
    return self.__page_titles.get(name)

  def search_post(self, word):
    return self.__posts.get(word)

File: ood/16_facebook.java
public enum ConnectionInvitationStatus{
  PENDING,
  ACCEPTED,
  REJECTED,
  CANCELED
}

public enum AccountStatus{
  ACTIVE,
  CLOSED,
  CANCELED,
  BLACKLISTED,
  DISABLED
}

public class Address {
  private String streetAddress;
  private String city;
  private String state;
  private String zipCode;
  private String country;
}




// For simplicity, we are not defining getter and setter functions. The reader can
// assume that all class attributes are private and accessed through their respective
// public getter method and modified only through their public setter method.

public class Account {
  private String id;
  private String password;
  private AccountStatus status;

  public boolean resetPassword();
}

public abstract class Person {
  private String name;
  private Address address;
  private String email;
  private String phone;

  private Account account;
}

public class Member extends Person {
  private Integer memberId;
  private Date dateOfMembership;
  private String name;

  private Profile profile;
  private HashSet<Integer> memberFollows;
  private HashSet<Integer> memberConnections;
  private HashSet<Integer> pageFollows;
  private HashSet<Integer> memberSuggestions;
  private HashSet<ConnectionInvitation> connectionInvitations;
  private HashSet<Integer> groupFollows;

  public boolean sendMessage(Message message);
  public boolean createPost(Post post);
  public boolean sendConnectionInvitation(ConnectionInvitation invitation);
  private Map<Integer, Integer> searchMemberSuggestions();
}

public class Admin extends Person {
  public boolean blockUser(Customer customer);
  public boolean unblockUser(Customer customer);
  public boolean enablePage(Page page);
  public boolean disablePage(Page page);
}

public class ConnectionInvitation {
  private Member memberInvited;
  private ConnectionInvitationStatus status;
  private Date dateCreated;
  private Date dateUpdated;

  public bool acceptConnection();
  public bool rejectConnection();
}







public class Profile {
  private byte[] profilePicture;
  private byte[] coverPhoto;
  private String gender;

  private List<Work> workExperiences;
  private List<Education> educations;
  private List<Place> places;
  private List<Stat> stats;

  public boolean addWorkExperience(Work work);
  public boolean addEducation(Education education);
  public boolean addPlace(Place place);
}

public class Work {
  private String title;
  private String company;
  private String location;
  private Date from;
  private Date to;
  private String description;
}






public class Page {
  private Integer pageId;
  private String name;
  private String description;
  private String type;
  private int totalMembers;
  private List<Recommendation> recommendation;

  private List<Recommendation> getRecommendation();
}

public class Recommendation {
  private Integer recommendationId;
  private int rating;
  private String description;
  private Date createdAt;
}







public class Group {
  private Integer groupId;
  private String name;
  private String description;
  private int totalMembers;
  private List<Member> members;

  public boolean addMember(Member member);
  public boolean updateDescription(String description);
}

public class Post {
  private Integer postId;
  private String text;
  private int totalLikes;
  private int totalShares;
  private Member owner;
}

public class Message {
  private Integer messageId;
  private Member[] sentTo;
  private String messageBody;
  private byte[] media;

  public boolean addMember(Member member);
}

public class Comment {
  private Integer commentId;
  private String text;
  private int totalLikes;
  private Member owner;
}







public interface Search {
  public List<Member> searchMember(String name);
  public List<Group> searchGroup(String name);
  public List<Page> searchPage(String name);
  public List<Post> searchPost(String word);
}

public class SearchIndex implements Search {
   HashMap<String, List<Member>> memberNames;
   HashMap<String, List<Group>> groupNames;
   HashMap<String, List<Page>> pageTitles;
   HashMap<String, List<Post>> posts;

   public boolean addMember(Member member) {
     if(memberNames.containsKey(member.getName())) {
       memberNames.get(member.getName()).add(member);
     } else {
       memberNames.put(member.getName(), member);
     }
   }

   public boolean addGroup(Group group);
   public boolean addPage(Page page);
   public boolean addPost(Post post);

  public List<Member> searchMember(String name) {
    return memberNames.get(name);
  }

  public List<Group> searchGroup(String name) {
    return groupNames.get(name);
  }

  public List<Page> searchPage(String name) {
    return pageTitles.get(name);
  }

  public List<Post> searchPost(String word) {
    return posts.get(word);
  }
}







import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.stream.Collectors;
import static java.util.Collections.reverseOrder;

public class Member extends Person {
  private Integer memberId;
  private Date dateOfMembership;
  private String name;

  private Profile profile;
  private HashSet<Integer> memberFollows;
  private HashSet<Integer> memberConnections;
  private HashSet<Integer> pageFollows;
  private HashSet<Integer> memberSuggestions;
  private HashSet<ConnectionInvitation> connectionInvitations;
  private HashSet<Integer> groupFollows;

  public boolean sendMessage(Message message);
  public boolean createPost(Post post);
  public boolean sendConnectionInvitation(ConnectionInvitation invitation);

  private Map<Integer, Integer> searchMemberSuggestions() {
    Map<Integer, Integer> suggestions = new HashMap<>();
    for(Integer memberId : this.memberConnections) {
      HashSet<Integer> firstLevelConnections = new Member(memberId).getMemberConnections());
      for(Integer firstLevelConnectionId : firstLevelConnections) {
        this.findMemberSuggestion(suggestions, firstLevelConnectionId);
        HashSet<Integer> secondLevelConnections = new Member(firstLevelConnectionId).getMemberConnections());
        for(Integer secondLevelConnectionId : secondLevelConnections) {
          this.findMemberSuggestion(suggestions, secondLevelConnectionId);
        }
      }
	  }

    // sort by value (increasing count), i.e., by highest number of mutual connection count
    Map<Integer, Integer> result = new LinkedHashMap<>();
    suggestions.entrySet().stream()
        .sorted(reverseOrder(Map.Entry.comparingByValue()))
        .forEachOrdered(x -> result.put(x.getKey(), x.getValue()));

    return result;
  }

  private void findMemberSuggestion(Map<Integer, Integer> suggestions, Integer connectionId) {
    // return if the proposed suggestion is already a connection or if there is a
    // pending connection invitation
    if(this.memberConnections.contains(connectionId) ||
        this.connectionInvitations.contains(connectionId)) {
      return;
    }

    int count = suggestions.containsKey(connectionId) ? suggestions.get(connectionId) : 0;
    suggestions.put(connectionId, count + 1);
  }
}





